source/renderer/app/stores/HardwareWalletsStore.ts

```diff
@@ -1786,6 +1585,388 @@ export default class HardwareWalletsStore extends Store {
       throw error;
     }
   };
+
+  @action
+  _findAssociatedWalletByExtendedPublicKey = async ({
+    extendedPublicKey,
+  }: FindAssociatedWalletByExtendedPublicKeyArgs): Promise<Wallet | null> => {
+    const deviceId =
+      extendedPublicKey?.deviceId || this.transportDevice.deviceId;
+    logger.debug('[HW-DEBUG] HWStore - EXPORT - deviceID: ', {
+      deviceId,
+    });
+
+    const hardwareWalletConnectionData = find(
+      this.hardwareWalletsConnectionData,
+      (hardwareWalletData) =>
+        extendedPublicKey.chainCodeHex ===
+          hardwareWalletData.extendedPublicKey.chainCodeHex &&
+        extendedPublicKey.publicKeyHex ===
+          hardwareWalletData.extendedPublicKey.publicKeyHex
+    );
+
+    return hardwareWalletConnectionData
+      ? this.stores.wallets.getWalletById(hardwareWalletConnectionData.id)
+      : null;
+  };
+
+  // Delete initiated (pending) device with this path since now is paired to wallet
+  @action
+  _deletePendingDeviceWithGivenPath = async ({
+    path,
+  }: DeletePendingDeviceWithGivenPathArgs) => {
+    const device = find(
+      this.hardwareWalletDevices,
+      // @ts-ignore ts-migrate(2339) FIXME: Property 'path' does not exist on type 'HardwareWa... Remove this comment to see the full error message
+      (device) => device.path === path
+    );
+
+    if (!device) return;
+
+    logger.debug(
+      '[HW-DEBUG] HWStore - __deletePendingDeviceWithGivenPath - UNSET Device with path: ',
+      {
+        deviceId: device.id,
+      }
+    );
+    await this._unsetHardwareWalletDevice({
+      deviceId: device.id,
+    });
+  };
+
+  @action
+  _discardConnectedDeviceAndReInitiateTransaction = ({
+    walletId,
+  }: DiscardConnectedDeviceAndReInitiateTransactionArgs = {}) => {
+    // Keep isTransactionInitiated active & Set new device listener by initiating transaction
+    // Show message to reconnect proper software wallet device pair
+    logger.debug(
+      '[HW-DEBUG] unfinishedWalletTxSigning SET: ',
+      // @ts-ignore ts-migrate(2345) FIXME: Argument of type 'string' is not assignable to par... Remove this comment to see the full error message
+      walletId
+    );
+    runInAction(
+      'HardwareWalletsStore:: set HW device CONNECTING FAILED',
+      () => {
+        this.hwDeviceStatus = HwDeviceStatuses.CONNECTING_FAILED;
```

Comment:
In this case, should we not also assign the `INCORRECT_PASSPHRASE_PROVIDED` status? (or like we discussed - UNRECOGNIZED_WALLET). If you check the implementation of the method which calls this method - the implementation and the comment suggest it is the same case:

```
if (!expectedWalletId || associatedWallet.id !== expectedWalletId) {
        logger.debug('[HW-DEBUG] HWStore - Device not belongs to this wallet');
        this._discardConnectedDeviceAndReInitiateTransaction({
          walletId: expectedWalletId,
        });
```

Reply:
That makes sense. The original issue was not connected to the scenario of the transaction so I didn't touch this case and during the refactoring just copied over the implementation. I will try to change the status to the one you suggested and reproduce that case to check it works fine :+1:

Reply:
That is also covered by my PR: https://github.com/input-output-hk/daedalus/pull/2860
I just noticed that because if we were to merge your refactoring as-is, my change would be reverted (assuming it will be merged first)